### PR TITLE
Feature/sg 492 fuzzy logic for get param and factories

### PR DESCRIPTION
**Pull request by shaydeci**
- Added support for typical typos (i.e non sensitivity of lowercase, uppercase and symbols in factories.
- Added factory unknown type exception, including a tip message regarding closest string match to they typo.
- get_param updated with the above fashion as well.
- metric_to_watch updated with the above fashion as well.
- Unit test: training with typos.

### CODE REVIEW

File: tests/unit_tests/factories_test.py [hunk 1]
```diff
@@ -38,6 +38,31 @@ def test_training_with_factories(self):
...
+            "valid_metrics_list": ["aCCuracy", "Top5"],  # use a metric by factory
+            "metric_to_watch": "Accurac_Y",
+            "greater_metric_to_watch_is_better": True,
+        }
+
+        trainer.train(model=net, training_params=train_params, train_loader=classification_test_dataloader(), valid_loader=classification_test_dataloader())
+
+        self.assertIsInstance(trainer.train_metrics.Accuracy, Accuracy)
+        self.assertIsInstance(trainer.valid_metrics.Top5, Top5)
+        self.assertIsInstance(trainer.optimizer, torch.optim.Adam)
```
Comment (ofrimasad) [inline tests/unit_tests/factories_test.py:64]: what about the loss function?

**Review by ofrimasad — COMMENTED** [review]
looks very nice. a few comments inline

File: src/super_gradients/common/exceptions/factory_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+     (out of 'choices' arg passed in __init__).
+
+    Attributes:
+        message -- explanation of the error
+    """
+
+    def __init__(self, unknown_type: str, choices: List, message: str = None):
+        message = message or f"Unknown object type: {unknown_type} in configuration. valid types are: {choices}"
+        if isinstance(unknown_type, str):
+            err_msg_tip = "\n Did you mean: " + str(process.extractOne(unknown_type, choices, scorer=fuzz.WRatio)[0]) + "?"
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/common/exceptions/factory_exceptions.py]: https://github.com/Deci-AI/super-gradients/pull/595/commits/aade8c821e741335053bfe335b107b8ce6423315 added.

File: src/super_gradients/training/utils/utils.py [hunk 3]
```diff
@@ -168,38 +168,80 @@ def tensor_container_to_device(obj: Union[torch.Tensor, tuple, list, dict], devi
         return obj
 
 
+def _fuzzy_keys(params: Mapping):
+    return [_fuzzy_str(s) for s in params.keys()]
+
+
+def _fuzzy_str(s):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/utils/utils.py]: true, didn't notice my uses. fixed in https://github.com/Deci-AI/super-gradients/pull/595/commits/1017a82014951b2e774777739b6bfca650e101b4

File: src/super_gradients/common/factories/type_factory.py [hunk 4]
```diff
@@ -32,6 +33,8 @@ def get(self, conf: Union[str, type]):
         if isinstance(conf, str) or isinstance(conf, bool):
             if conf in self.type_dict:
                 return self.type_dict[conf]
+            elif isinstance(conf, str) and get_param(self.type_dict, conf) is not None:
+                return get_param(self.type_dict, conf)
             else:
                 try:
                     lib = ".".join(conf.split(".")[:-1])
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/common/factories/type_factory.py:41]: Added in the designate area below https://github.com/Deci-AI/super-gradients/pull/595/commits/8f8ab2bd4280f4ff4392b9146e6fc2a35ae1d93c .

Raising it before is not desired as we also want to support dot paths (I.e "torch.optim.Adam")

File: tests/unit_tests/factories_test.py (hunk 1 above)
Comment (shaydeci) [reply tests/unit_tests/factories_test.py:65]: Added in latest commit.

**Review by ofrimasad — APPROVED** [review]
LGTM
